### PR TITLE
132 serial numbers postgres

### LINKED ISSUE #298: Serial Numbers
We seem to be getting some duplicates, not sure if it's data file issue or sync issue, but it's much easier to reason about numbers (invoice, requisition, stocktake) by just using max number for type. Since we have a hard sync rule about record ownership, this shouldn't be an issue.

- [ ] Remove number table entirely
- [ ] Just use max serial number (by type) when creating new records

### LINKED ISSUE #442: Avoid conflicts in serial number
From: 

https://github.com/openmsupply/open-msupply/pull/423

@clemens-msupply 

Just thought a bit about race conditions. Sqlite has a transaction isolation level of SERIALIZABLE which means if a tx succeeds its like a single user accessed the DB -> all good. Postgres has a lower isolation level of READ COMMITTED which can lead to tx anomalies.

Already in the old code there was a bug in next_number. If two transaction run concurrently they could both start with same old number which would result in a number used twice. However, this can be fixed by using SELECT FOR UPDATE (see https://www.postgresql.org/docs/current/explicit-locking.html row level locks).

With the new way to calculate the number I am not sure if this would work, e.g. don't think it helps to lock the row with the highest number because two tx might have a different idea what this row is.

We can increase the isolation level for the particular tx but this would be quite complicated and would require us to handle retries if there is a concurrent number update.

We might need to lock the table in ACCESS EXCLUSIVE mode? I haven't worked with advisory locks but this might be a solution as well. @jmbrunskill do you have some knowledge about it?

@jmbrunskill 

I haven't read all the code or understand the full situation, but here's some thoughts.

1) Databases like postgres provide a `sequence` generator for purposes like this, they are multi user safe so we could generate the next invoice number using this. https://www.postgresql.org/docs/9.1/functions-sequence.html
It looks like postgres allows this sequence to be set to a particular number, so would allow us to say set it to the largest number seen after a sync or something if that's a requirement.

I remember Chris saying something about number-reuse where we fill in numbers that didn't get used? 
I hope this isn't still a requirement with open-mSupply?

2) Could we just use BigSerial for our invoice numbers? I haven't tried it but I'm pretty sure you can insert into a serial column without breaking the serial (e.g. for syncing the records.) Sqlite has autoincrement which would serve a similar purpose. To use this, if you have other records that need that value,  you'll need to get the serial number back from the db after an insert.

3) I agree an Exclusive lock or advisory lock is probably required if we want to ensure two concurrent queries can't get the same 'max' record. 

>  SELECT FOR UPDATE will wait for a concurrent transaction that has run any of those commands on the same row, and will then lock and return the updated row (or no row, if the row was deleted). 
https://www.postgresql.org/docs/current/explicit-locking.html

My read of this suggests both SELECT FOR UPDATE queries would return the same row and then increment to receive the  same number. Advisory locks look like they would probably work here, we could create one for each `next_number` table. It would allow pure read transactions to succeed on the table, but prevent reads that hold that lock from running concurrently.

We'd need to make sure the lock is acquired at the higher level transaction where the record is written, not just in the 'get max number' transaction, otherwise we could end up not fixing the problem at all.

I think a `Sequence` or `Serial` are probably the best solutions unless we have some business rules or constraint that would mean we need something more custom.

@jmbrunskill 

After reading the code a bit more I see the goal is to give a incrementing number for each store and type, so can't just use a serial or auto increment if we need that constraint.

Creating a sequence for each use case it probably plausible, but might be hard to reason about and require raw sql, as not sure how it would map on to diesel. 

@jmbrunskill 

Advisory Locks sound like a good solution for postgres.
I guess we'd need to create a function that we call to set and unset the lock that is a no-op on sqlite. We wouldn't want to have postgres specific code leaking too deeply into the code base.

Pretty sure you can just start the transaction with this
`SELECT pg_advisory_xact_lock(123);` 

This should unlock when the transaction ends, which sounds like the safest way to do this.
We could probably use an enum for the lock_id (123 in this example), so we only lock the related record type (e.g 123 == invoice 124 == requisition).

Alternatively, we could probably have a number service in rust that gets initialised from the database on startup, and ensures non-concurrent access via a mutex, but I think that's unnecessarily complex.

@clemens-msupply 

Or just have one global Mutex serializing any number generation. Would avoid sqlite/postgres differences (and at least for me easier to understand, I am never 100% sure what is locked/safe in Postgres)

@jmbrunskill 

> Or just have one global Mutex serializing any number generation. Would avoid sqlite/postgres differences (and at least for me easier to understand, I am never 100% sure what is locked/safe in Postgres)

Agreed, we'd just need to hold the mutex till the database write is completed, or decouple the number generation from the database `max` query.

What would happen during a sync? I think it will be ok because we shouldn't be generating numbers for our stores on any other site right?

@andreievg 

> What would happen during a sync? I think it will be ok because we shouldn't be generating numbers for our stores on any other site right?

Yep

> Or just have one global Mutex serializing any number generation. Would avoid sqlite/postgres differences (and at least for me easier to understand, I am never 100% sure what is locked/safe in Postgres)

I think this would work nicely as long as we grab mutex before transaction starts, services are kind of meant to be `mixable`, as in transaction can be already ongoing and another service may call a service with connection in a transaction, so I would say it might be hard.

We kind of need a quick mutex increment that can happen within a transaction and will alway give a new number. Potentially we would want to keep number table for this.

### A

* On startup grab max for all types and initialise number provider
* Number provider is shared across all services
* When number is requested, lock and provide next number (incrementing cached next number)

### B

Actually, i can't think of any other way because of these constraints:
* When invoice/requisition service is called it could already be in a transaction
* Can't start write transaction while another write transaction is ongoing in Sqlite

### Suggestion

I think we want to start with benchmark / test since we would need this anyways to validate that our changes work, I think we can look at the threshold of when this problem occurs, and access the urgency of the fix (bench can be done as a test, but it would required having two async runtimes I think).

### LINKED ISSUE #468: Number table: custom number column
Programs needs some custom numbering. For example, a number for HIV programs.

- [ ] introduce a NumberRowType::Custom enum variant
- [ ] add a `custom_name: Option<String>`
- [ ] extent the number service to get a next number for a custom number

TODO move to another issue:
- [ ] add om_custom_name column to mSupply (4d)
- [ ] sync new column 

Also see: https://github.com/openmsupply/open-msupply/issues/442#issuecomment-1205795906

### PR DESCRIPTION
**Pull request by jmbrunskill**
Closes #298
Closes #468 
Closes #442 

- Test for concurrent 'next_number' requests for both insert (First number) and update (subsequent numbers) cases
-  The fix is to do most of the work in the repository layer relying on the 'returning' feature from postgres and sqlite to return the data from any updated or inserted rows.
- Add unique constraint to db

Things get a little funky in the insert case as postgres rolls back the transaction if there is a unique constraint violation without `ON CONFLICT DO NOTHING` which isn't supported in sqlite.

### CODE REVIEW

**Comment by jmbrunskill** [comment]
Thought I should add a comment about why I selected a different solution to what was discussed.

1. I didn't like the extra round trips to the database, getting the current value, incrementing it in rust and updating the record. It seems inefficient and more risk of a concurrency error and more load on the database.
2. Thought it would result in the same code for postgres and sqlite however, I've found that it's can't quite be the same (although possible could be if diesel supported 'returning' on sqlite, it doesn't seem to be implemented in our current version which is why we have the raw sql strings)

Happy to go back to the postgres select for update option if we think it would be better for some reason.

File: server/repository/src/db_diesel/number_row.rs [hunk 1]
```diff
@@ -55,6 +88,68 @@ impl<'a> NumberRowRepository<'a> {
...
+                    ON_CONFLICT_DO_NOTHING
+                );
+
+                let insert_query = sql_query(insert_query_str).bind::<Text, _>(store_id);
+                let insert_result =
+                    insert_query.get_result::<NextNumber>(&self.connection.connection);
+
+                match insert_result {
+                    Ok(result) => Ok(result),
+                    Err(NotFound) => {
```
Comment (clemens-msupply) [inline server/repository/src/db_diesel/number_row.rs:134]: Cool this initial number insert was ignored before.

File: server/repository/src/db_diesel/number_row.rs [hunk 2]
```diff
@@ -55,6 +88,68 @@ impl<'a> NumberRowRepository<'a> {
...
+
+                        match update_query.get_result::<NextNumber>(&self.connection.connection) {
+                            Ok(result) => Ok(result),
+                            Err(e) => Err(RepositoryError::from(e)),
+                        }
+                    }
+                    Err(e) => Err(RepositoryError::from(e)),
+                }
+            }
+            Err(e) => Err(RepositoryError::from(e)),
```
Comment (clemens-msupply) [inline server/repository/src/db_diesel/number_row.rs:149]: Do you want to add a test case for all these 3 cases?

File: server/repository/src/db_diesel/number_row.rs [hunk 3]
```diff
@@ -55,6 +88,68 @@ impl<'a> NumberRowRepository<'a> {
...
+                    Err(NotFound) => {
+                        // 3. If we got here another thread inserted the record before we we able to (we know this because nothing was returned for the insert)
+                        // We should now be able to do the same 'update returning' query as before to get our new number.
+
+                        let update_query =
+                            sql_query(update_query_str.clone()).bind::<Text, _>(store_id);
+
+                        match update_query.get_result::<NextNumber>(&self.connection.connection) {
+                            Ok(result) => Ok(result),
+                            Err(e) => Err(RepositoryError::from(e)),
```
Comment (clemens-msupply) [inline server/repository/src/db_diesel/number_row.rs:143]: I think this match can be replace by just:

```rust
let result = update_query.get_result::<NextNumber>(&self.connection.connection)?;
Ok(result)
```

File: server/repository/migrations/postgres/2021-12-10T10-00_number_table/up.sql [hunk 4]
```diff
@@ -1,17 +1,20 @@
...
     id TEXT NOT NULL PRIMARY KEY,
     -- current counter value
     value BIGINT NOT NULL,
     store_id TEXT NOT NULL REFERENCES store(id),
     type number_type NOT NULL
-)
\ No newline at end of file
+  );
+
+CREATE UNIQUE INDEX ix_number_store_type_unique ON number(store_id, type);
```
Comment (clemens-msupply) [inline server/repository/migrations/postgres/2021-12-10T10-00_number_table/up.sql:20]: Is this for performance or for uniqueness? Might be bad for performance because there are only a handful of rows and I can imagine a full table scan is faster...
Or is it for "ON CONFLICT DO NOTHING" to work?

**Review by clemens-msupply — APPROVED** [review]
Great thanks! As you mentioned bit ugly to use raw sql and more complex, but on the plus side we now handle the initialization edge case if a number row does not exist and two processes request the next number at the same time.

File: server/service/src/number.rs [hunk 5]
```diff
@@ -75,4 +59,133 @@ mod test {
...
+            let result: Result<i64, TransactionError<RepositoryError>> = connection
+                .transaction_sync(|con| {
+                    let num = next_number(con, &NumberRowType::Stocktake, "store_a")?;
+                    std::thread::sleep(core::time::Duration::from_millis(TEST_SLEEP_TIME));
+                    Ok(num)
+                });
+            result.unwrap()
+        });
+
+        //Part 2: Concurrent up date both doing updates
```
[the remark this answers is not in the document]
Comment (jmbrunskill) [reply server/service/src/number.rs:123]: Part 1 should result in an insert and an update. Part 2 should result in 2 updates. Maybe wording should be improved here.

File: server/repository/src/db_diesel/number_row.rs (hunk 2 above)
Comment (jmbrunskill) [reply server/repository/src/db_diesel/number_row.rs:149]: I believe the test case provided tests all 3 cases. Let me know if you think something is missed.

File: server/repository/migrations/postgres/2021-12-10T10-00_number_table/up.sql (hunk 4 above)
Comment (jmbrunskill) [reply server/repository/migrations/postgres/2021-12-10T10-00_number_table/up.sql:20]: This is for uniqueness. Hopefully the database engine will still pick a tablescan if it's the fastest option.

File: server/service/src/number.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply server/service/src/number.rs:123]: I added some debug output and the "insert new row conflict" (second error) is not reached. However, difficult to trigger in a test since you would need to pause in the repository method...

File: server/repository/src/db_diesel/number_row.rs (hunk 2 above)
Comment (clemens-msupply) [reply server/repository/src/db_diesel/number_row.rs:149]: I added some debug output and the "insert new row conflict" (second error) is not reached. However, difficult to trigger in a test since you would need to pause in the repository method...

Comment (jmbrunskill) [reply server/repository/src/db_diesel/number_row.rs:149]: Interesting it's definitely being hit sometimes in my test, are you testing with postgres or sqlite? SQL doesn't have this issue due to the single write transaction guarantee.

Comment (clemens-msupply) [reply server/repository/src/db_diesel/number_row.rs:149]: tested with postgres...

Comment (jmbrunskill) [reply server/repository/src/db_diesel/number_row.rs:149]:
```
PART 1
PART 2
PART 1
PART 2
PART 3
next_number (INSERT) results : a=1 b=2
PART 1
PART 1
next_number (UPDATE) results : a=3 b=4
test number::test::test_concurrent_next_number ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 151 filtered out; finished in 0.81s
```

Here's what I get on my M1 Mac. Maybe it's because my postgres is a bit slow in docker?

File: server/repository/src/db_diesel/number_row.rs (hunk 3 above)
Comment (jmbrunskill) [reply server/repository/src/db_diesel/number_row.rs]: Yes, that's better thanks!
